### PR TITLE
Improve and refine MLP tests for extensibility and A/B testing

### PR DESCRIPTION
**Pull request by rpsilva-aws**
In this PR, we include various fixes, improvements and extensions, namely:
- Exposing the MLP test to other tests (to allow us to A/B test convergences: requirement for the grad acc tests)
- Improving the asserts, and extending to losses and outputs
- Using the appropriate flag for the LR and log steps
- Improving the model layout with nn.Sequential
- Enhance the coverage, by actually utilizing the checkpointing flag, and including a sanity test for CPU
- Decouple to simplify the A/B coverage
- Fix imports

### CODE REVIEW

**Comment by rpsilva-aws** [comment]
@tengyifei @ManfeiBai I found myself having to largely improve/enhance the MLP tests, since I wanted to reuse this test for A/B convergence validation with/without gradient accumulation (using experimental grad acc loops) and with the same batch sizes. PTAL.

File: test/neuron/run_tests.sh [hunk 1]
```diff
@@ -213,7 +213,7 @@ function run_xla_op_tests3 {
   #run_test "$CDIR/spmd/test_dtensor_integration2.py"
   run_test "$CDIR/spmd/test_xla_auto_sharding.py"
   #run_test "$CDIR/spmd/test_spmd_parameter_wrapping.py"
-  run_test "$CDIR/spmd/test_train_spmd_linear_model.py"
```
Comment (tengyifei) [inline test/neuron/run_tests.sh:216]: Did you intend to remove this test?

File: test/run_tests.sh [hunk 2]
```diff
@@ -231,6 +231,7 @@ function run_xla_op_tests3 {
   run_test "$CDIR/spmd/test_xla_auto_sharding.py"
   run_test "$CDIR/spmd/test_spmd_parameter_wrapping.py"
   run_test "$CDIR/spmd/test_mp_input_sharding.py"
+  run_test "$CDIR/spmd/test_train_spmd_linear_model.py"
```
Comment (tengyifei) [inline test/run_tests.sh:234]: Should we have `test_train_spmd_linear_model_grad_checkpointing` here?

Should we have `test_train_spmd_linear_model_grad_checkpointing` in TPU tests as well?

**Review by tengyifei — APPROVED** [review]

File: test/neuron/run_tests.sh (hunk 1 above)
Comment (rpsilva-aws) [reply test/neuron/run_tests.sh:216]: Yes, it's basically a subset of the checkpoint one. We cross check their convergence, while ensuring that the losses/output is not 0. We only have `test_train_spmd_linear_model` for CPU as standalone, since it doesn't run with grad checkpointing there.

File: test/run_tests.sh (hunk 2 above)
Comment (rpsilva-aws) [reply test/run_tests.sh:234]: See above. CPU is the only case we use `test_train_spmd_linear_model` instead of `test_train_spmd_linear_model_grad_checkpointing`. I didn't see much value in having `test_train_spmd_linear_model` as standalone in addition to the checkpoint one. Perhaps I could rephrase the test file name, since it runs baseline as part of its A/B assertions.

**Review by tengyifei — CHANGES_REQUESTED** [review]
Synced offline. Let's add a `--skip-gradient-checkpointing` CLI arg to the train testing script or similar to skip the gradient checkpointing on CPU, in order to avoid the confusing `test_train_spmd_linear_model_grad_checkpointing` name.

**Comment by rpsilva-aws** [comment]
> Synced offline. Let's add a `--skip-gradient-checkpointing` CLI arg to the train testing script or similar to skip the gradient checkpointing on CPU, in order to avoid the confusing `test_train_spmd_linear_model_grad_checkpointing` name.

I was just about to write this, done!

File: test/run_tests.sh [hunk 3]
```diff
@@ -231,6 +231,7 @@ function run_xla_op_tests3 {
   run_test "$CDIR/spmd/test_xla_auto_sharding.py"
   run_test "$CDIR/spmd/test_spmd_parameter_wrapping.py"
   run_test "$CDIR/spmd/test_mp_input_sharding.py"
+  run_test "$CDIR/spmd/test_train_spmd_linear_model.py" "$@" --skip-gradient-checkpointing
```
Comment (tengyifei) [inline test/run_tests.sh:234]: Could you add this test to `test/tpu/run_tests.sh` (and i presume neuron) too?

Comment (rpsilva-aws) [reply test/run_tests.sh:234]: It was already on those, and now neuron and TPU will run with gradient checkpointing as an additional coverage.

Comment (rpsilva-aws) [reply test/run_tests.sh:234]: - https://github.com/pytorch/xla/blob/master/test/tpu/run_tests.sh#L17
- https://github.com/pytorch/xla/blob/master/test/neuron/run_tests.sh#L217

Comment (tengyifei) [reply test/run_tests.sh:234]: ack, thanks

**Review by tengyifei — APPROVED** [review]

File: test/utils/train_spmd_linear_model.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
+import sys
```
Comment (lsy323) [inline test/utils/train_spmd_linear_model.py:1]: qq is this just a duplicated test file?

**Review by lsy323 — CHANGES_REQUESTED** [review]
Looks like the test file is duplicated under `test/utils`?

File: test/utils/train_spmd_linear_model.py (hunk 4 above)
Comment (rpsilva-aws) [reply test/utils/train_spmd_linear_model.py:1]: This was mostly the former `test_train_spmd_linear_model.py`, minus all the different improvements (you can see the different commits). There is no duplication in the code with this PR.

**Comment by rpsilva-aws** [comment]
> Looks like the test file is duplicated under `test/utils`?

I don't see it. The core of the functionality was moved to utils - it's a standalone training script. The `test_` file can run simultaneous executions of it, to, intentionally, run A/B testing.

**Review by lsy323 — APPROVED** [review]

**Comment by lsy323** [comment]
> > Looks like the test file is duplicated under `test/utils`?
> 
> I don't see it. The core of the functionality was moved to utils - it's a standalone training script. The `test_` file can run simultaneous executions of it, to, intentionally, run A/B testing. The test/utils file is not for direct testing, but users can run it as a standalone script, and so can tests.

Hi @rpsilva-aws thank you for the explanation. I still don't get the A/B testing part. But if the body can be reused in other tests then it's fine

**Comment by rpsilva-aws** [comment]
> > > Looks like the test file is duplicated under `test/utils`?
> > 
> > 
> > I don't see it. The core of the functionality was moved to utils - it's a standalone training script. The `test_` file can run simultaneous executions of it, to, intentionally, run A/B testing. The test/utils file is not for direct testing, but users can run it as a standalone script, and so can tests.
> 
> Hi @rpsilva-aws thank you for the explanation. I still don't get the A/B testing part. But if the body can be reused in other tests then it's fine

Thanks. You can see it in this PR here: https://github.com/pytorch/xla/pull/8561/files#diff-09c6a280d1c6fc8053d5a17e919e29fe51a75fd593f6c2012496b6e970312c25R44

Essentially, it allows it to A/B test different functionality. It adds a bit more value than running each standalone test, and just checking that the losses/output are not 0. Alternatively, we'd need to assert against an hardcoded set of expected values. We want to reuse this for testing gradient accumulation with and without XLA's while loop.
